Fixes #417: Apply weather instantly when teleporting

lgromanowski opened the pull request:
Change speed of weather transition from blight to other (twice fast as normal)
and from other to blight (four times faster than normal).

Signed-off-by: Lukasz Gromanowski lgromanowski@gmail.com

lgromanowski committed:
Fixes #417: Apply weather instantly when teleporting

Change speed of weather transition from blight to other (twice fast as normal)
and from other to blight (four times faster than normal).

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

ghost commented:
Using arbitrary speed multipliers seems like a bad idea. I think we should try to detect teleports to another cell and don't have any transition in that case.

ghost commented:
Perhaps track the player position in the weather manager and if the player moved by a large distance in one frame, discard transitions?

lgromanowski commented:
I think tracking should be done in WorldImp and when teleportation is detected then it should call ie. WeatherManager::switchToNextWeather(bool instantly = true);

lgromanowski committed:
Fixes #417: Apply weather instantly when teleporting

Removed changing speed of weather transition introduced in previous
commit. Instead try to detect player "teleporting" (ie. coc),
and then switch instantly to the next weather type.

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

zinnschlag commented:
Storing the old player position in a static variable isn't going to work. It can cause all kind of problems (e.g. when loading a saved game or when taking interiors into consideration). It also will totally blow up, once we have multiple worldspaces.

btw. we us all-upercase names exclusively for preprocessor constants.

lgromanowski commented:
I think it shouldn't be a problem at all, because WeatherManager::update() and WeatherManager::switchToNextWeather() checks if current cell is exterior or interior. That static player position is only used to check if player was moved for larger amount of units than predefined constant in one frame, so we can detect coc, or any kind of "teleporting", it shouldn't affect normal walking/running/flying (I've tested it with a npc with normal and boosted stats/attributes). I will change all-uprecase constant to camelcase.

lgromanowski committed:
Fixes #417: Apply weather instantly when teleporting

Corrected constant name.

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

zinnschlag commented:
Are we sure that the exterior/interior issue is covered? Scenario: We teleport from an interior cell with weather to the exact same position in an exterior cell.
Also, there is still the multiple worldspace issue. Little point in using an implementation now that will have to be ripped out in a few month (multiple worldspaces is one of the things I intent to do first after 1.0).

lgromanowski commented:
You're right. What do you think about storing a cell in additional variable and at first checking if cell changed and then compare x,y position?

zinnschlag commented:
Wow, this issue is tricky.

I think we are on the wrong track here. Implicitly detecting a teleport is error-prone and just a waste of time. Instead all events that cause a teleport should set a flag in MWWorld::Player. And then on a weather update the flag should be checked (and cleared). If the flag was set and a region change has happened, the weather change should be instantaneous.
This may be a bit more work initially, but at least it will function reliably and we don't have to worry about any edge cases.

Causes of a teleport I can think of:
- Script instructions (setPos, Position, PositionCell)
- Teleport magic effects (recall and the temple teleport spells): Not entirely sure if that is implemented.
- Player going to prison after a crime: Not implemented yet.
- Using a teleport door

lgromanowski committed:
Fixes #417: Apply weather instantly when teleporting

Changed teleporting detection from "position tracking" to manually
setting "teleportation" flag ( player->setTeleported(true) ).

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

lgromanowski commented:
All causes of teleport you mentioned, except when player is going to prison are handled in last commit.

lgromanowski committed:
Fixes #417: Apply weather instantly when teleporting

Correction to previous commit - WeatherManager->update()
will always be called.

Signed-off-by: Lukasz Gromanowski <lgromanowski@gmail.com>

lgromanowski commented on apps/openmw/mwworld/worldimp.cpp:
```diff
@@ -2244,4 +2244,17 @@
+        if (mPlayer->wasTeleported())
+        {
+            mPlayer->setTeleported(false);
+            mWeatherManager->switchToNextWeather(true);
```
ok, corrected

lgromanowski commented on apps/openmw/mwworld/weather.cpp:
```diff
@@ -324,42 +324,7 @@ void WeatherManager::update(float duration)
-
-        setWeather(weatherType, false);
-    }
+    switchToNextWeather(false);
```
Sorry, I'll correct this in a few moments.
